Fixes #26698: Refactoring of ValidationNeeded class : replace type Box with IOResult

skaerg opened the pull request:
https://issues.rudder.io/issues/26698

skaerg commented:
Commit modified

skaerg commented:
Commit modified

skaerg commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJdbcRepository.scala:
```diff
@@ -291,29 +299,24 @@ class WoChangeRequestJdbcRepository(
 
     for {
-      cr      <- roRepo.get(changeRequest.id)
-      ok      <- cr match {
-                   case None    =>
-                     val msg = s"Cannot update non-existent Change Request with id ${changeRequest.id.value}"
-                     ChangeValidationLogger.warn(msg)
-                     Failure(msg)
-                   case Some(x) => Full("ok")
-                 }
-      update  <- {
+      _       <- roRepo
+                   .get(changeRequest.id)
+                   .notOptional(s"Cannot update non-existent Change Request with id ${changeRequest.id.value}")
+                   .tapError(err => ChangeValidationLoggerPure.error(err.fullMsg))
+      _       <- {
```
replaced the call to `ChangeValidationLogger.warn` with `ChangeValidationLoggerPure.error` since the block returns a Failure (instead of a Full accompanied by a warning). do tell me whether i should change it back

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJdbcRepository.scala:
```diff
@@ -154,8 +154,8 @@ class RoChangeRequestJdbcRepository(
   // utility method which correctly transform Doobie types towards Box[Vector[ChangeRequest]]
-  private[this] def execQuery(q: Query0[Box[ChangeRequest]]): Box[Vector[ChangeRequest]] = {
-    transactRunBox(xa => {
+  private[this] def execQuery(q: Query0[Box[ChangeRequest]]): IOResult[Vector[ChangeRequest]] = {
+    transactIOResult("conversion from type Doobie to type Box[Vector[ChangeRequest]] has failed")(xa => {
```
Since `execQuery` is a method to factorize an operation of `transactIOResult` on a query that may fail, I think the message should always be specified by the caller of `execQuery`, if not the error cause will be too generic and the final error message will end up being confusing. 

So the error message should be as a parameter of the `execQuery` method, and that would also require that every database query/operation have explicit error message, which will improve the consistency and user-friendliness.
For example : 
* `getAll()` should fail with an error message `"Could not select all change request in database"`
* `getByDirective(id, onlyPending)` should fail with an error message `"Could not get change requests for directive ${id}"`

You could define it like so : 
```suggestion
  private[this] def execQuery(errorMsg: String)(q: Query0[Box[ChangeRequest]]): IOResult[Vector[ChangeRequest]] = {
    transactIOResult(errorMsg)(xa => {
```

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJdbcRepository.scala:
```diff
@@ -291,29 +299,24 @@ class WoChangeRequestJdbcRepository(
+      _       <- {
         val (name, desc, xml, modId) = getAtom(changeRequest)
-        transactRunBox(xa => updateChangeRequestSQL(name, desc, xml, modId, changeRequest.id).run.transact(xa))
+        transactIOResult(s"Could not update the change request with id ${changeRequest.id} in database")(xa =>
```
:+1: that is a clear error message

clarktsiory reviewed (changes requested):
There are some error messages that need to be more explicit

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/ChangeRequestJdbcRepository.scala:
```diff
@@ -259,19 +264,18 @@ class WoChangeRequestJdbcRepository(
-                Failure(msg)
-              case Some(x) => Full(x)
-            }
+      id <- transactIOResult(s"Could not create change request with id ${changeRequest.id}")(xa =>
```
in database ?

clarktsiory reviewed (changes requested):
Yet some error that could be more explicit by using `transactIOResult` : we should indicate that we are working with the database, because operations such a "create a change request" could occur at one or more level (API, cache, database, file, etc.)

skaerg commented:
PR updated with a new commit

skaerg commented:
Commit modified

skaerg commented:
Commit modified

skaerg commented:
PR updated with a new commit

skaerg commented:
4th commit : 
- corrected line 109 in `ChangeValidationConf` for which the translation from Box to IOResult was erroneous
- added an errMsg parameter to `execQuery` in `ChangeRequestJdbcRepository`, and updated the error messages which are passed to `execQuery`

NOTE : There are actually a few other files, apart from the snippets, that still use the `Box` type in `change-validation` which i failed to notice last week : 
- ChangeValidationPluginDef
- WorkflowJdbcRepositoryTest
- ChangeRequestJdbcRepository
- ChangeRequestJdbcRepositoryTest
- ChangeRequestJson

clarktsiory reviewed (changes requested):
* There are some failing database tests, but it is normal that they were not caught because they need to be run by hand...
* Also there are some remaining effects that can be grouped to be run once
* Some message that need to be adapted


Other than that, the changes look great and usage of ZIO is idiomatic, and the error-handling is a lot clearer, GG !

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/snippet/ChangeRequestChangesForm.scala:
```diff
@@ -108,30 +109,34 @@ class ChangeRequestChangesForm(
+            ruleCategoryRepository
+              .getRootCategory()
+              .chainError("An error occurred when trying to get data from base. ")
+              .fold(
+                err => {
+                  logger.error(err.fullMsg)
+                  Text(err.fullMsg)
+                },
+                rootRuleCategory => {
+                  ("#changeTree ul *" #> new ChangesTreeNode(cr, rootRuleCategory).toXml &
+                  "#history *" #> displayHistory(
```
this `fold` embeds some Lift code into the ZIO context

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/snippet/ChangeRequestDetails.scala:
```diff
@@ -412,17 +422,23 @@ implicit private val qc: QueryContext = CurrentUser.queryContext // bug https://
+        val (_, evalNextStep) = nextChosen
+        evalNextStep(cr.id, CurrentUser.actor, changeMessage.map(_.get))
+          .chainError("could not change Change request step")
+          .fold(
+            err => {
+              formTracker.addFormError(error(err.fullMsg))
+              logger.error(s"Error when saving change request '${cr.id.value}': ${err.fullMsg}")
+              updateForm(nextChosen)
+            },
+            next => {
+              SetHtml("workflowActionButtons", displayActionButton(cr, next)) &
+              SetHtml("newStatus", Text(next.value)) &
+              closePopup & JsRaw(""" initBsModal("successWorkflow"); """) // JsRaw ok, const
+            }
+          )
+          .runNow
```
this `fold` embeds some Lift code into the ZIO context

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/comet/WorkflowInformation.scala:
```diff
@@ -78,32 +78,34 @@ class WorkflowInformation extends CometActor with CometListener with Loggable {
+      .fold(
+        err => {
+          logger.error(err.fullMsg)
+          (".dropdown-menu *" #> <li class="dropdown-header">{err.fullMsg}</li>).apply(layout)
+        },
+        workflowEnabled => {
+          val cssSelect = {
+            if (workflowEnabled && (isValidator || isDeployer)) {
+              {
+                if (isValidator) pendingModifications
+                else ".dropdown-menu *+" #> NodeSeq.Empty
+              } & {
+                if (isDeployer) pendingDeployment
+                else ".dropdown-menu *+" #> NodeSeq.Empty
+              } &
+              "#number *" #> requestCount(workflowService)
+            } else {
+              ".dropdown *" #> NodeSeq.Empty
+            }
           }
+          cssSelect(layout)
         }
-
-        cssSelect(layout)
-    }
+      )
+      .runNow
```
this `fold` embeds some Lift code into the ZIO context

clarktsiory reviewed (changes requested):
Some `fold` are not supported by Lift because it embeds some Lift code in the success part. It throws : 
```
INFO  net.liftweb.http.LiftRules - Snippet Failure: SnippetFailure(/secure/configurationManager/changes/changeRequest/1 -> ParsePath(List(secure, configurationManager, changes, changeRequest),,true,false),Full(ChangeRequestDetails.changes),Access to Lift's statefull features from Stateless mode)
```

The solution is to use `.either.runNow match { case Left(...) / case Right(...) }` instead

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/snippet/ChangeRequestDetails.scala:
```diff
@@ -412,17 +422,23 @@ implicit private val qc: QueryContext = CurrentUser.queryContext // bug https://
-        }
+
+        val (_, evalNextStep) = nextChosen
+        evalNextStep(cr.id, CurrentUser.actor, changeMessage.map(_.get))
```
The `CurrentUser.actor` should be initialized into a variable at the top of this method, to avoid issues, as in https://github.com/Normation/rudder/pull/6280

skaerg commented:
PR updated with a new commit

skaerg commented:
PR rebased

skaerg commented:
Commit modified

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/WorkflowJdbcRepository.scala:
```diff
@@ -37,30 +37,30 @@
 
-  def getStateOfChangeRequest(crId: ChangeRequestId): Box[WorkflowNodeId]
+  def getStateOfChangeRequest(crId: ChangeRequestId): IOResult[WorkflowNodeId]
 
-  def getAllChangeRequestsState(): Box[Map[ChangeRequestId, WorkflowNodeId]]
+  def getAllChangeRequestsState(): IOResult[Map[ChangeRequestId, WorkflowNodeId]]
 }
 
 trait WoWorkflowRepository {
-  def createWorkflow(crId: ChangeRequestId, state: WorkflowNodeId): Box[WorkflowNodeId]
+  def createWorkflow(crId: ChangeRequestId, state: WorkflowNodeId): IOResult[WorkflowNodeId]
 
-  def updateState(crId: ChangeRequestId, from: WorkflowNodeId, state: WorkflowNodeId): Box[WorkflowNodeId]
+  def updateState(crId: ChangeRequestId, from: WorkflowNodeId, state: WorkflowNodeId): IOResult[WorkflowNodeId]
```
jdbc tests needs to be corrected

skaerg commented:
PR updated with a new commit

skaerg commented:
PR updated with a new commit
make changes in PR review

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/ValidationNeeded.scala:
```diff
@@ -109,22 +107,25 @@ class NodeGroupValidationNeeded(
+      targets          = Set(change.newRule) ++ change.previousRule.toSet
+      res              = checkNodeTargetByRule(groups, arePolicyServer.mapValues(_.rudderSettings.isPolicyServer), supervised, targets)
+      end             <- com.normation.zio.currentTimeMillis
+      _                = {
+        ChangeValidationLogger.Metrics.debug(
+          s"Check rule '${change.newRule.name}' [${change.newRule.id.serialize}]" +
+          s"change requestion need for validation in ${end - start}ms"
+        )
+      }
```
```suggestion
      _                <- {
        ChangeValidationLoggerPure.Metrics.debug(
          s"Check rule '${change.newRule.name}' [${change.newRule.id.serialize}]" +
          s"change requestion need for validation in ${end - start}ms"
        )
      }
```

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/ValidationNeeded.scala:
```diff
@@ -161,69 +162,76 @@ class NodeGroupValidationNeeded(
+      exists      = groups
+                      .getNodeIds(supervised.map(identity), nodeFacts.mapValues(_.rudderSettings.isPolicyServer))
+                      .find(nodeId => targetNodes.contains(nodeId))
+      res         = {
+        // we want to let the log know why the change request needs validation
+        exists.foreach { nodeId =>
+          ChangeValidationLogger.debug(
+            s"Node '${nodeId.value}' belongs to both a supervised group and to group '${change.newGroup.name}' [${change.newGroup.id.serialize}]"
+          )
+        }
+        exists.nonEmpty
+      }
+      end        <- com.normation.zio.currentTimeMillis
+      _           = {
+        ChangeValidationLogger.Metrics.debug(
+          s"Check group '${change.newGroup.name}' [${change.newGroup.id.serialize}] " +
+          s"change requestion need for validation in ${end - start}ms"
```
```suggestion
      res        <- {
        // we want to let the log know why the change request needs validation
        ZIO.foreach(exists) { nodeId =>
          ChangeValidationLoggerPure.debug(
            s"Node '${nodeId.value}' belongs to both a supervised group and to group '${change.newGroup.name}' [${change.newGroup.id.serialize}]"
          ).map(_.nonEmpty)
        }
      }
      end        <- com.normation.zio.currentTimeMillis
      _          <- {
        ChangeValidationLoggerPure.Metrics.debug(
          s"Check group '${change.newGroup.name}' [${change.newGroup.id.serialize}] " +
          s"change requestion need for validation in ${end - start}ms"
```

clarktsiory commented on change-validation/src/main/scala/com/normation/plugins/changevalidation/ValidationNeeded.scala:
```diff
@@ -161,69 +162,76 @@ class NodeGroupValidationNeeded(
+      res         =
+        checkNodeTargetByRule(groups, nodeFacts.mapValues(_.rudderSettings.isPolicyServer), supervised, (rules ++ newRules).toSet)
+      end        <- com.normation.zio.currentTimeMillis
+      _           = {
+        ChangeValidationLogger.Metrics.debug(
```
```suggestion
      _           <- {
        ChangeValidationLoggerPure.Metrics.debug(
```

clarktsiory reviewed (commented):
A few last suggestions, loggers can be pure ones

skaerg commented:
Commit modified

skaerg commented:
Commit modified

skaerg commented:
Commit modified

skaerg commented:
PR rebased

clarktsiory reviewed (approved):
LGTM ! Tested with the counterpart https://github.com/Normation/rudder/pull/6320, it the code is now very idiomatic and functional, and effects seem to be handled correctly at boundaries !

Normation-Quality-Assistant commented:
OK, merging this PR
